app/src/main/java/com/m2049r/xmrwallet/data/BarcodeData.java

```diff
@@ -175,13 +180,55 @@ static public BarcodeData fromString(String qrCode) {
             // maybe it's naked?
             bcData = parseNaked(qrCode);
         }
+        if (bcData == null) {
+            // Check for UD domain
+            bcData = parseUD(qrCode);
+        }
         if (bcData == null) {
             // check for OpenAlias
             bcData = parseOpenAlias(qrCode, false);
         }
         return bcData;
     }
 
+    static public BarcodeData parseUD(String udString) {
+        Timber.d("parseUD=%s", udString);
+        if (udString == null) return null;
+
+        DomainResolution resolution = Resolution.builder()
+                .providerUrl(NamingServiceType.ENS, "https://cloudflare-eth.com")
+                .build();
+
+        final String[] address = {null};
+        final Crypto[] crypto = {null};
+        Thread udThread = new Thread(() -> {
+            for (Crypto currentCrypto: Crypto.values()) {
+                try {
+                    address[0] = resolution.getAddress(udString, currentCrypto.getSymbol().toLowerCase());
+                    crypto[0] = currentCrypto;
+                    break;
+                } catch (NamingServiceException e) {
+                    Timber.d(e.getLocalizedMessage());
+                }
+            }
+        });
+        udThread.start();
+        try {
+            udThread.join();
```

Comment:
why start a thread and then block the UI thread waiting for it to finish anyway?